src/move/emojicoin_arena/sources/emojicoin_arena.move

```diff
@@ -0,0 +1,703 @@
+        let (melee_just_ended, _registry_ref_mut, _time, _n_melees_before_cranking) =
+            crank_schedule();
+        exit_inner<Coin0, LP0, Coin1, LP1>(participant, melee_id, !melee_just_ended);
+    }
+
+    #[randomness]
+    entry fun swap<Coin0, LP0, Coin1, LP1>(
+        swapper: &signer, melee_id: u64, market_addresses: vector<address>
+    ) acquires MeleeEscrow, Registry, UserMelees {
+        let (exit_once_done, _registry_ref_mut, _time, _n_melees_before_cranking) =
+            crank_schedule();
+
+        // Return early if type arguments or melee ID are passed incorrectly, but only after
+        // cranking schedule.
+        let swapper_address = signer::address_of(swapper);
+        if (!exists<MeleeEscrow<Coin0, LP0, Coin1, LP1>>(swapper_address)) {
+            return;
+        };
+        let escrow_ref_mut = &mut MeleeEscrow<Coin0, LP0, Coin1, LP1>[swapper_address];
+        if (escrow_ref_mut.melee_id != melee_id)
+            return;
+
+        if (coin::value(&escrow_ref_mut.emojicoin_0) > 0) {
+            swap_within_escrow<Coin0, LP0, Coin1, LP1>(
+                swapper,
+                swapper_address,
+                market_addresses[0],
+                market_addresses[1],
+                &mut escrow_ref_mut.emojicoin_0,
+                &mut escrow_ref_mut.emojicoin_1
+            );
+        } else {
+            swap_within_escrow<Coin1, LP1, Coin0, LP0>(
+                swapper,
+                swapper_address,
+                market_addresses[1],
+                market_addresses[0],
+                &mut escrow_ref_mut.emojicoin_1,
+                &mut escrow_ref_mut.emojicoin_0
+            );
+        };
+
+        if (exit_once_done) exit_inner<Coin0, LP0, Coin1, LP1>(swapper, melee_id, false);
+    }
+
+    fun init_module(arena: &signer) acquires Registry {
+        // Store registry resource.
+        let (vault_signer, signer_capability) =
+            account::create_resource_account(arena, REGISTRY_SEED);
+        move_to(
+            arena,
+            Registry {
+                melees_by_id: smart_table::new(),
+                melee_ids_by_market_ids: smart_table::new(),
+                signer_capability,
+                new_melee_duration: DEFAULT_DURATION,
+                new_melee_available_rewards: DEFAULT_AVAILABLE_REWARDS,
+                new_melee_max_match_percentage: DEFAULT_MAX_MATCH_PERCENTAGE,
+                new_melee_max_match_amount: DEFAULT_MAX_MATCH_AMOUNT,
+                all_entrants: smart_table::new(),
+                n_melee_swaps: aggregator_v2::create_unbounded_aggregator(),
+                melee_swaps_volume: aggregator_v2::create_unbounded_aggregator(),
+                rewards_disbursed: 0
+            }
+        );
+        coin::register<AptosCoin>(&vault_signer);
+
+        // Use pseudo-randomness to get market IDs for the first melee, since randomness is not
+        // supported during `init_module`.
```

Comment:
> Is there a reason we don't just have a wrapper for the crank function to init the first melee after package initialization, in a context where true randomness is available ?

Because it is simplest to let publication trigger the whole process, and right now the first crank is a special case that would lead to more complex logic in the general cranking mechanism. Note that even though the pseudo-randomness is technically biasable in the general case, for the purposes of package initialization it is not going to be subect to a test-and-repeat approach in practice, because the underlying source of pseudo-randomness traces back to `aptos_framework::transaction_context::generate_unique_address` which states:

>  Returns a universally unique identifier (of type address) generated
    /// by hashing the transaction hash of this transaction and a sequence number
    /// specific to this transaction. 

E.g. this would only be in practice biased if the publisher were to simulate and retry transactions based on the simulated output (a legitimate concern for public APIs which rely on randomness), but the publisher has no such incentive, e.g. random in practice

Reply:
@CRBl69 please resolve if no further comments